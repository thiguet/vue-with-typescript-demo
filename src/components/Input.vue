<template>
    <fieldset>
        <legend>
            <span class="accessibility">{{ legend  }}</span>
        </legend>
        <label :for="id" class="label">{{ label }}</label>
        <input  :id="id"
                :type="type"
                :name="id"
                class="input"
                v-bind:value="value"
                v-on:input="setValue" />
    </fieldset>
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component({
  name: 'Input',
})
export default class Input extends Vue {
    @Prop({ required: true }) readonly id!: string;

    @Prop({ required: true }) readonly label!: string;

    @Prop({ required: true }) readonly value!: string;

    @Prop({ required: true }) readonly setValue!: Function;

    @Prop({ default: 'text' }) readonly type!: string;

    @Prop() readonly legend!: string;
}
</script>

<style scoped>
fieldset {
    display: flex;
    flex-direction: row;
    flex-flow: row;
    border: none;
    width: 100%;
}

input {
    border: none;
    border-bottom: 1px solid var(--border-color);
    width: 100%;
    font-family: inherit;
    font-size: var(--font-input-size);
}

label {
    margin-bottom: 5px;
}

label:after {
    content: ':';
}

legend span.accessibility {
    position:absolute;
    left:-9999px;
    width:100px;
    height:auto;
    overflow:hidden;
}
</style>
