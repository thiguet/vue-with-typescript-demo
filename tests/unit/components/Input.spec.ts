import { shallowMount } from '@vue/test-utils';
import Input from '@/components/Input.vue';
import faker from 'faker';

interface InputData {
  id: string;
  label: string;
  value: string;
  setValue: Function;
  legend?: string;
  type?: "text" | "number" | "password" | undefined;
};

describe('Input.vue', () => {
  let props: InputData;
  
  const build = () => {
    const wrapper = shallowMount(Input, {
      propsData: { ...props },
    });

    return {
      wrapper,
      input: () => wrapper.find('.input'),
      labelEl: () => wrapper.find('.label'),
      legend: () => wrapper.find('legend'),
    };
  };

  beforeEach(() => {
    props = {
      id: faker.random.alphaNumeric(),
      type: 'text',
      value: faker.random.word(),
      setValue: jest.fn(),
      label: faker.random.word(),
      legend: "This is an Input",
    };
  });

  it('renders component', () => {
    props.id = 'someFixedValidId';
    props.value = 'someFixedValidValue';
    props.label = 'someFixedValidLabel';

    const { wrapper } = build();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders main child components', () => {
    const {
      input,
      labelEl,
    } = build();

    expect(input().exists()).toBe(true);
    expect(labelEl().exists()).toBe(true);
  });

  it('passes props to the component', () => {
    const { wrapper } = build();

    const {
      type,
      value,
      label,
      id,
      setValue,
      legend,
    } = wrapper.props();

    expect(id).toBe(props.id);
    expect(type).toBe(props.type);
    expect(value).toBe(props.value);
    expect(label).toBe(props.label);
    expect(setValue).toBe(props.setValue);
    expect(legend).toBe(props.legend);
  });

  it('sets label propety to label element', () => {
    const {
      wrapper,
      labelEl,
    } = build();

    const { label } = wrapper.props();

    expect(labelEl().text()).toBe(label);
  });

  it('sets input\'s type propety to element by props', () => {
    const {
      wrapper,
      input,
    } = build();

    const { type } = wrapper.props();

    expect(input().attributes().type).toBe(type);
  });

  it('sets type to text using default value', () => {
    delete props.type;

    const { wrapper } = build();

    expect(wrapper.props().type).toBe("text");
  });

  it('has legend element', () => {
    const {
      wrapper,
      legend,
    } = build();

    expect(legend().text()).toBe(wrapper.props().legend);
  });

  it('must have label for input element', () => {
    const {
      wrapper,
      labelEl,
      input,
    } = build();

    expect(labelEl().attributes().for).toBe(props.id);
    expect(labelEl().attributes().for).toBe(input().attributes().name);
  });
});
