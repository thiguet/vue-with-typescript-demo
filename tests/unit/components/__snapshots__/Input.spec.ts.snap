// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.vue renders component 1`] = `
<fieldset>
  <legend><span class="accessibility">This is an Input</span></legend> <label for="someFixedValidId" class="label">someFixedValidLabel</label> <input id="someFixedValidId" type="text" name="someFixedValidId" class="input">
</fieldset>
`;
